File: tetris.py
import pygame
import random
# This is the original, single file version of the game.
# It is now obsolete.
# It has been refactored into several modules that are imported into main.py.
# So, to launch the game, enter the command python main.py

pygame.init()

# Constants
SCREEN_WIDTH, SCREEN_HEIGHT = 400, 500
BLOCK_SIZE = 25
GRID_WIDTH, GRID_HEIGHT = SCREEN_WIDTH // BLOCK_SIZE, SCREEN_HEIGHT // BLOCK_SIZE

# Colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 0, 0)
CYAN = (0, 255, 255)
YELLOW = (255, 255, 0)
MAGENTA = (255, 0, 255)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)

# Shapes
# Each shape is represented as a list of lists, where 1 indicates a filled block and 0 indicates an empty space
SHAPES = [
    # I Shape (Straight)
    [[1, 1, 1, 1]],
    
    # O Shape (Square)
    [[1, 1],
     [1, 1]],
    
    # T Shape (Tee)
    [[1, 1, 1],
     [0, 1, 0]],
    
    # J Shape (Left L)
    [[1, 1, 1],
     [1, 0, 0]],
    
    # L Shape (Right L)
    [[1, 1, 1],
     [0, 0, 1]],
    
    # S Shape (Left S)
    [[1, 1, 0],
     [0, 1, 1]],
    
    # Z Shape (Right Z)
    [[0, 1, 1],
     [1, 1, 0]]
]

# Initialize game
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("Tetris")
clock = pygame.time.Clock()

# Function to draw the grid lines
def draw_grid(grid):
    for x in range(0, SCREEN_WIDTH, BLOCK_SIZE):
        pygame.draw.line(screen, WHITE, (x, 0), (x, SCREEN_HEIGHT))
    for y in range(0, SCREEN_HEIGHT, BLOCK_SIZE):
        pygame.draw.line(screen, WHITE, (0, y), (SCREEN_WIDTH, y))

    for y, row in enumerate(grid):
        for x, (filled, color) in enumerate(row):
            if filled:
                draw_block(x, y, color)

# Function to draw a single block
def draw_block(x, y, color):
    pygame.draw.rect(screen, color, (x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE))

# Function to rotate a shape clockwise
def rotate_shape(shape):
    return list(zip(*reversed(shape)))

# Function to place a shape on the grid
def place_shape_on_grid(grid, shape, x, y, color):
    for dy, row in enumerate(shape):
        for dx, value in enumerate(row):
            if value:
                grid[y + dy][x + dx] = (1, color)

# Function to check for collision with the walls or other blocks
def check_collision(grid, shape, x, y):
    for dy, row in enumerate(shape):
        for dx, value in enumerate(row):
            if value and (x + dx < 0 or x + dx >= GRID_WIDTH or y + dy >= GRID_HEIGHT or grid[y + dy][x + dx][0]):
                return True
    return False

# Function to check if the game is over
def check_game_over(grid):
    return any(grid[0][x][0] for x in range(GRID_WIDTH))

def main():
    # Initialize the grid with black color
    grid = [[(0, BLACK) for _ in range(GRID_WIDTH)] for _ in range(GRID_HEIGHT)]
    
    # Select a random shape and color for the current falling shape
    current_shape = random.choice(SHAPES)
    current_color = random.choice([RED, CYAN, YELLOW, MAGENTA, GREEN, BLUE])
    
    # Initial position of the falling shape
    shape_x = GRID_WIDTH // 2 - len(current_shape[0]) // 2
    shape_y = 0

    game_over = False

        # Timing variables
    normal_tick_rate = 500  # milliseconds for regular descent
    fast_tick_rate = 100    # milliseconds for faster descent
    last_tick = pygame.time.get_ticks()

    horizontal_move_delay = 150  # milliseconds delay for horizontal movement
    last_horizontal_move = pygame.time.get_ticks()

    rotation_delay = 200  # milliseconds delay for rotation
    last_rotation_time = pygame.time.get_ticks()

    while not game_over:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()

        keys = pygame.key.get_pressed()

        # Handle horizontal movement with delay
        current_time = pygame.time.get_ticks()
        if current_time - last_horizontal_move > horizontal_move_delay:
            if keys[pygame.K_LEFT] and shape_x > 0 and not check_collision(grid, current_shape, shape_x - 1, shape_y):
                shape_x -= 1
                last_horizontal_move = current_time
            if keys[pygame.K_RIGHT] and shape_x + len(current_shape[0]) < GRID_WIDTH and not check_collision(grid, current_shape, shape_x + 1, shape_y):
                shape_x += 1
                last_horizontal_move = current_time

        # Handle rotation with delay
        if keys[pygame.K_UP] and current_time - last_rotation_time > rotation_delay:
            rotated_shape = rotate_shape(current_shape)
            if shape_x + len(rotated_shape[0]) <= GRID_WIDTH and shape_y + len(rotated_shape) <= GRID_HEIGHT \
                    and not check_collision(grid, rotated_shape, shape_x, shape_y):
                current_shape = rotated_shape
                last_rotation_time = current_time

        # Determine the descent rate
        current_tick_rate = fast_tick_rate if keys[pygame.K_DOWN] else normal_tick_rate

        # Handle automatic downward movement
        if current_time - last_tick > current_tick_rate:
            if not check_collision(grid, current_shape, shape_x, shape_y + 1):
                shape_y += 1
            else:
                place_shape_on_grid(grid, current_shape, shape_x, shape_y, current_color)
                if check_game_over(grid):
                    game_over = True
                else:
                    shape_y = 0
                    shape_x = GRID_WIDTH // 2 - len(current_shape[0]) // 2
                    current_shape = random.choice(SHAPES)
                    current_color = random.choice([RED, CYAN, YELLOW, MAGENTA, GREEN, BLUE])
            last_tick = current_time

        screen.fill(BLACK)
        draw_grid(grid)

        # Draw current shape
        for y, row in enumerate(current_shape):
            for x, value in enumerate(row):
                if value:
                    draw_block(shape_x + x, shape_y + y, current_color)

        pygame.display.flip()
        clock.tick(30)

    # Display "Game Over" message
    font = pygame.font.Font(None, 36)
    game_over_text = font.render("Game Over", True, WHITE)
    screen.blit(game_over_text, (SCREEN_WIDTH // 2 - 100, SCREEN_HEIGHT // 2 - 18))
    pygame.display.flip()
    pygame.time.wait(2000)

if __name__ == "__main__":
    main()

File: tetris.js
// // Constants
// const canvas = document.getElementById("gameCanvas");
// const ctx = canvas.getContext("2d");
// const SCREEN_WIDTH = 400;
// const SCREEN_HEIGHT = 500;
// const BLOCK_SIZE = 25;
// const GRID_WIDTH = SCREEN_WIDTH / BLOCK_SIZE;
// const GRID_HEIGHT = SCREEN_HEIGHT / BLOCK_SIZE;
// canvas.width = SCREEN_WIDTH;
// canvas.height = SCREEN_HEIGHT;

// // Colors and Shapes
// const COLORS = {
//     0: "#00FFFF", // Cyan (I)
//     1: "#FFFF00", // Yellow (O)
//     2: "#FF00FF", // Magenta (T)
//     3: "#0000FF", // Blue (J)
//     4: "#FFA500", // Orange (L)
//     5: "#00FF00", // Green (S)
//     6: "#FF0000", // Red (Z)
//     empty: "#000000", // Black (grid background)
// };
// const SHAPES = [
//     [[1, 1, 1, 1]], // I
//     [[1, 1], [1, 1]], // O
//     [[1, 1, 1], [0, 1, 0]], // T
//     [[1, 1, 1], [0, 0, 1]], // J
//     [[1, 1, 1], [1, 0, 0]], // L
//     [[0, 1, 1], [1, 1, 0]], // S
//     [[1, 1, 0], [0, 1, 1]], // Z
// ];

// // Game Variables
// let grid, currentShape, currentColor, shapeX, shapeY, nextShape, nextColor;
// let score = 0, level = 0, linesCleared = 0;
// let gameRunning = false, gamePaused = false;
// let lastTick = 0, normalTickRate = 1000, fastTickRate = 100;
// let isFastDropping = false; // Tracks if down arrow is held

// // Initialize Grid
// function initializeGrid() {
//     return Array.from({ length: GRID_HEIGHT }, () => Array(GRID_WIDTH).fill(0));
// }

// // Draw Grid
// function drawGrid() {
//     ctx.fillStyle = COLORS.empty;
//     ctx.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
//     for (let y = 0; y < GRID_HEIGHT; y++) {
//         for (let x = 0; x < GRID_WIDTH; x++) {
//             if (grid[y][x]) {
//                 ctx.fillStyle = COLORS[grid[y][x]];
//                 ctx.fillRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
//                 ctx.strokeStyle = "#FFF";
//                 ctx.strokeRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
//             }
//         }
//     }
// }

// // Draw Shape
// function drawShape(shape, color, offsetX, offsetY) {
//     ctx.fillStyle = color;
//     for (let y = 0; y < shape.length; y++) {
//         for (let x = 0; x < shape[y].length; x++) {
//             if (shape[y][x]) {
//                 ctx.fillRect(
//                     (shapeX + x + offsetX) * BLOCK_SIZE,
//                     (shapeY + y + offsetY) * BLOCK_SIZE,
//                     BLOCK_SIZE,
//                     BLOCK_SIZE
//                 );
//                 ctx.strokeStyle = "#FFF";
//                 ctx.strokeRect(
//                     (shapeX + x + offsetX) * BLOCK_SIZE,
//                     (shapeY + y + offsetY) * BLOCK_SIZE,
//                     BLOCK_SIZE,
//                     BLOCK_SIZE
//                 );
//             }
//         }
//     }
// }

// // Rotate Shape
// function rotateShape(shape) {
//     return shape[0].map((_, colIndex) => shape.map(row => row[colIndex]).reverse());
// }

// // Check Collision
// function checkCollision(shape, offsetX, offsetY) {
//     for (let y = 0; y < shape.length; y++) {
//         for (let x = 0; x < shape[y].length; x++) {
//             if (
//                 shape[y][x] &&
//                 (shapeY + y + offsetY >= GRID_HEIGHT || // Bottom boundary
//                     shapeX + x + offsetX < 0 || // Left boundary
//                     shapeX + x + offsetX >= GRID_WIDTH || // Right boundary
//                     grid[shapeY + y + offsetY]?.[shapeX + x + offsetX]) // Occupied space
//             ) {
//                 return true;
//             }
//         }
//     }
//     return false;
// }

// // Place Shape on Grid
// function placeShapeOnGrid(shape, color) {
//     for (let y = 0; y < shape.length; y++) {
//         for (let x = 0; x < shape[y].length; x++) {
//             if (shape[y][x]) {
//                 grid[shapeY + y][shapeX + x] = Object.keys(COLORS).find(key => COLORS[key] === color);
//             }
//         }
//     }
// }

// // Clear Full Rows
// function clearFullRows() {
//     let clearedRows = 0;
//     for (let y = GRID_HEIGHT - 1; y >= 0; y--) {
//         if (grid[y].every(cell => cell !== 0)) {
//             grid.splice(y, 1);
//             grid.unshift(Array(GRID_WIDTH).fill(0));
//             clearedRows++;
//             y++;
//         }
//     }
//     return clearedRows;
// }

// // Calculate Score
// function calculateScore(lines, level) {
//     const lineScores = { 1: 40, 2: 100, 3: 300, 4: 1200 };
//     return (lineScores[lines] || 0) * (level + 1);
// }

// // Spawn New Shape
// function spawnShape() {
//     const shapeIndex = Math.floor(Math.random() * SHAPES.length);
//     currentShape = SHAPES[shapeIndex];
//     currentColor = COLORS[shapeIndex];
//     nextShape = SHAPES[Math.floor(Math.random() * SHAPES.length)];
//     nextColor = COLORS[Object.keys(COLORS).find(key => COLORS[key] === nextShape)];
//     shapeX = Math.floor((GRID_WIDTH - currentShape[0].length) / 2);
//     shapeY = 0;

//     if (checkCollision(currentShape, 0, 0)) {
//         gameRunning = false; // Game over
//     }
// }

// // Game Loop
// function gameLoop(timestamp) {
//     if (!gameRunning || gamePaused) return;

//     const tickRate = isFastDropping ? fastTickRate : normalTickRate;
//     if (timestamp - lastTick >= tickRate) {
//         if (!checkCollision(currentShape, 0, 1)) {
//             shapeY++;
//         } else {
//             placeShapeOnGrid(currentShape, currentColor);
//             const clearedRows = clearFullRows();
//             if (clearedRows > 0) {
//                 score += calculateScore(clearedRows, level);
//                 linesCleared += clearedRows;
//                 if (linesCleared % 10 === 0) level++;
//             }
//             spawnShape();
//         }
//         lastTick = timestamp;
//     }

//     drawGrid();
//     drawShape(currentShape, currentColor, 0, 0);
//     requestAnimationFrame(gameLoop);
// }

// // Start Game
// document.getElementById("startButton").addEventListener("click", () => {
//     grid = initializeGrid();
//     spawnShape();
//     gameRunning = true;
//     gamePaused = false;
//     score = 0;
//     level = 0;
//     linesCleared = 0;
//     requestAnimationFrame(gameLoop);
// });

// // Pause Game
// document.getElementById("pauseButton").addEventListener("click", () => {
//     gamePaused = !gamePaused;
//     if (!gamePaused) requestAnimationFrame(gameLoop);
// });

// // Handle Keyboard Input
// document.addEventListener("keydown", (event) => {
//     if (!gameRunning || gamePaused) return;

//     if (event.key === "ArrowLeft" && !checkCollision(currentShape, -1, 0)) {
//         shapeX--;
//     } else if (event.key === "ArrowRight" && !checkCollision(currentShape, 1, 0)) {
//         shapeX++;
//     } else if (event.key === "ArrowDown") {
//         isFastDropping = true;
//     } else if (event.key === "ArrowUp") {
//         const rotatedShape = rotateShape(currentShape);
//         if (!checkCollision(rotatedShape, 0, 0)) {
//             currentShape = rotatedShape;
//         }
//     }
// });

// // Handle Key Release for Down Arrow
// document.addEventListener("keyup", (event) => {
//     if (event.key === "ArrowDown") {
//         isFastDropping = false;
//     }
// });

// Constants
const canvas = document.getElementById("gameCanvas");
const ctx = canvas.getContext("2d");
const SCREEN_WIDTH = 400;
const SCREEN_HEIGHT = 500;
const BLOCK_SIZE = 25;
const GRID_WIDTH = SCREEN_WIDTH / BLOCK_SIZE;
const GRID_HEIGHT = SCREEN_HEIGHT / BLOCK_SIZE;
canvas.width = SCREEN_WIDTH;
canvas.height = SCREEN_HEIGHT;

// Colors and Shapes
const COLORS = {
    0: "#00FFFF", // Cyan (I)
    1: "#FFFF00", // Yellow (O)
    2: "#FF00FF", // Magenta (T)
    3: "#0000FF", // Blue (J)
    4: "#FFA500", // Orange (L)
    5: "#00FF00", // Green (S)
    6: "#FF0000", // Red (Z)
    empty: "#000000", // Black (grid background)
};
const SHAPES = [
    [[1, 1, 1, 1]], // I
    [[1, 1], [1, 1]], // O
    [[1, 1, 1], [0, 1, 0]], // T
    [[1, 1, 1], [0, 0, 1]], // J
    [[1, 1, 1], [1, 0, 0]], // L
    [[0, 1, 1], [1, 1, 0]], // S
    [[1, 1, 0], [0, 1, 1]], // Z
];

// Game Variables
let grid, currentShape, currentColor, shapeX, shapeY, nextShape, nextColor;
let score = 0, level = 0, linesCleared = 0;
let gameRunning = false, gamePaused = false;
let lastTick = 0, normalTickRate = 1000, fastTickRate = 100;
let isFastDropping = false; // Tracks if down arrow is held

// DOM Elements
const scoreElement = document.getElementById("score");
const levelElement = document.getElementById("level");
const linesElement = document.getElementById("lines");

// Initialize Grid
function initializeGrid() {
    return Array.from({ length: GRID_HEIGHT }, () => Array(GRID_WIDTH).fill(0));
}

// Draw Grid
function drawGrid() {
    ctx.fillStyle = COLORS.empty;
    ctx.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
    for (let y = 0; y < GRID_HEIGHT; y++) {
        for (let x = 0; x < GRID_WIDTH; x++) {
            if (grid[y][x]) {
                ctx.fillStyle = COLORS[grid[y][x]];
                ctx.fillRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
                ctx.strokeStyle = "#FFF";
                ctx.strokeRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
            }
        }
    }
}

// Draw Shape
function drawShape(shape, color, offsetX, offsetY) {
    ctx.fillStyle = color;
    for (let y = 0; y < shape.length; y++) {
        for (let x = 0; x < shape[y].length; x++) {
            if (shape[y][x]) {
                ctx.fillRect(
                    (shapeX + x + offsetX) * BLOCK_SIZE,
                    (shapeY + y + offsetY) * BLOCK_SIZE,
                    BLOCK_SIZE,
                    BLOCK_SIZE
                );
                ctx.strokeStyle = "#FFF";
                ctx.strokeRect(
                    (shapeX + x + offsetX) * BLOCK_SIZE,
                    (shapeY + y + offsetY) * BLOCK_SIZE,
                    BLOCK_SIZE,
                    BLOCK_SIZE
                );
            }
        }
    }
}

// Rotate Shape
function rotateShape(shape) {
    return shape[0].map((_, colIndex) => shape.map(row => row[colIndex]).reverse());
}

// Check Collision
function checkCollision(shape, offsetX, offsetY) {
    for (let y = 0; y < shape.length; y++) {
        for (let x = 0; x < shape[y].length; x++) {
            if (
                shape[y][x] &&
                (shapeY + y + offsetY >= GRID_HEIGHT || // Bottom boundary
                    shapeX + x + offsetX < 0 || // Left boundary
                    shapeX + x + offsetX >= GRID_WIDTH || // Right boundary
                    grid[shapeY + y + offsetY]?.[shapeX + x + offsetX]) // Occupied space
            ) {
                return true;
            }
        }
    }
    return false;
}

// Place Shape on Grid
function placeShapeOnGrid(shape, color) {
    for (let y = 0; y < shape.length; y++) {
        for (let x = 0; x < shape[y].length; x++) {
            if (shape[y][x]) {
                grid[shapeY + y][shapeX + x] = Object.keys(COLORS).find(key => COLORS[key] === color);
            }
        }
    }
}

// Clear Full Rows
function clearFullRows() {
    let clearedRows = 0;
    for (let y = GRID_HEIGHT - 1; y >= 0; y--) {
        if (grid[y].every(cell => cell !== 0)) {
            grid.splice(y, 1);
            grid.unshift(Array(GRID_WIDTH).fill(0));
            clearedRows++;
            y++;
        }
    }
    return clearedRows;
}

// Calculate Score
function calculateScore(lines, level) {
    const lineScores = { 1: 40, 2: 100, 3: 300, 4: 1200 };
    return (lineScores[lines] || 0) * (level + 1);
}

// Update UI
function updateUI() {
    scoreElement.textContent = score;
    levelElement.textContent = level;
    linesElement.textContent = linesCleared;
}

// Spawn New Shape
function spawnShape() {
    const shapeIndex = Math.floor(Math.random() * SHAPES.length);
    currentShape = SHAPES[shapeIndex];
    currentColor = COLORS[shapeIndex];
    nextShape = SHAPES[Math.floor(Math.random() * SHAPES.length)];
    nextColor = COLORS[Object.keys(COLORS).find(key => COLORS[key] === nextShape)];
    shapeX = Math.floor((GRID_WIDTH - currentShape[0].length) / 2);
    shapeY = 0;

    if (checkCollision(currentShape, 0, 0)) {
        gameRunning = false; // Game over
    }
}

// Game Loop
function gameLoop(timestamp) {
    if (!gameRunning || gamePaused) return;

    const tickRate = isFastDropping ? fastTickRate : normalTickRate;
    if (timestamp - lastTick >= tickRate) {
        if (!checkCollision(currentShape, 0, 1)) {
            shapeY++;
        } else {
            placeShapeOnGrid(currentShape, currentColor);
            const clearedRows = clearFullRows();
            if (clearedRows > 0) {
                score += calculateScore(clearedRows, level);
                linesCleared += clearedRows;
                if (linesCleared % 10 === 0) level++;
                updateUI(); // Update UI after scoring
            }
            spawnShape();
        }
        lastTick = timestamp;
    }

    drawGrid();
    drawShape(currentShape, currentColor, 0, 0);
    requestAnimationFrame(gameLoop);
}

// Start Game
document.getElementById("startButton").addEventListener("click", () => {
    grid = initializeGrid();
    spawnShape();
    gameRunning = true;
    gamePaused = false;
    score = 0;
    level = 0;
    linesCleared = 0;
    updateUI(); // Reset UI
    requestAnimationFrame(gameLoop);
});

// Pause Game
document.getElementById("pauseButton").addEventListener("click", () => {
    gamePaused = !gamePaused;
    if (!gamePaused) requestAnimationFrame(gameLoop);
});

// Handle Keyboard Input
document.addEventListener("keydown", (event) => {
    if (!gameRunning || gamePaused) return;

    if (event.key === "ArrowLeft" && !checkCollision(currentShape, -1, 0)) {
        shapeX--;
    } else if (event.key === "ArrowRight" && !checkCollision(currentShape, 1, 0)) {
        shapeX++;
    } else if (event.key === "ArrowDown") {
        isFastDropping = true;
    } else if (event.key === "ArrowUp") {
        const rotatedShape = rotateShape(currentShape);
        if (!checkCollision(rotatedShape, 0, 0)) {
            currentShape = rotatedShape;
        }
    }
});

// Handle Key Release for Down Arrow
document.addEventListener("keyup", (event) => {
    if (event.key === "ArrowDown") {
        isFastDropping = false;
    }
});

// DOM Elements for new buttons
const rotateButton = document.getElementById("rotateButton");
const speedupButton = document.getElementById("speedupButton");
const leftButton = document.getElementById("leftButton");
const rightButton = document.getElementById("rightButton");

// Handle Button Controls
rotateButton.addEventListener("click", () => {
    if (gameRunning && !gamePaused) {
        const rotatedShape = rotateShape(currentShape);
        if (!checkCollision(rotatedShape, 0, 0)) {
            currentShape = rotatedShape;
        }
    }
});

speedupButton.addEventListener("mousedown", () => {
    if (gameRunning && !gamePaused) {
        isFastDropping = true;
    }
});

speedupButton.addEventListener("mouseup", () => {
    isFastDropping = false;
});

leftButton.addEventListener("click", () => {
    if (gameRunning && !gamePaused && !checkCollision(currentShape, -1, 0)) {
        shapeX--;
    }
});

rightButton.addEventListener("click", () => {
    if (gameRunning && !gamePaused && !checkCollision(currentShape, 1, 0)) {
        shapeX++;
    }
});
